### PR TITLE
Support DictionaryString for Regex matching operators

### PR DESCRIPTION
**Pull request by blaginin**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/12618

## Rationale for this change

As explained in the original PR, regex comparison operations don't support dictionaries like `~*`.

## What changes are included in this PR?

When building a logical query, the query is initially correctly coerced to `Utf8` (which is supported in `BinaryExpr`), but then `unwrap_cast_in_comparison` _removes_ the column cast. 

image

Interestingly, for the `Operator::LikeMatch` operator, which is similar to regex operators, `unwrap_cast_in_comparison` doesn't rewrite the query. This happens because of the function currently named `is_comparison_operator`.

According to its usages, it is only used in interval propagation, and the number of actual arguments supported is actually [smaller](https://github.com/apache/datafusion/blob/cf76aba36d56f775a61bb434da5e8e3b07d9061d/datafusion/physical-expr/src/intervals/cp_solver.rs#L317-L327) than the number mentioned in the method.

Moreover, the name `is_comparison_operator` seems a bit confusing, as it doesn't match the list of comparison operators in the [docs](https://github.com/apache/datafusion/blob/cf76aba36d56f775a61bb434da5e8e3b07d9061d/docs/source/user-guide/sql/operators.md?plain=1#L105). Therefore, I have renamed the method.

## Are these changes tested?

Yes, I uncommented tests in `datafusion/sqllogictest/test_files/string`.

## Are there any user-facing changes?

No

### CODE REVIEW

**Comment by alamb** [comment]
Thank you for this contribution @blaginin  -- I agree this PR seems to have fixed the issue, but I am a little worried about it.

1. It is probably slower ( I think this PR effectively leaves the cast to Utf8 there (which would basically unpack the dictionary array))
2. It seems to change interval analysis and propagation to not support `LIKE` etc.


FYI @ozankabak  I am a bit worried that the changes to the propagation logic did not cause any test failures 🤔  Maybe we need to increase test coverage in that area.

**Comment by alamb** [comment]
What would you think about explicitly implementing physical operator support for DictionaryArrays? That would mean basically applying the operation on the [`DictionaryArray::value`](https://docs.rs/arrow/latest/arrow/array/struct.DictionaryArray.html#method.values)s 

That would look something like 

```rust
let input_array: DictionaryArray<Int32, String> = ...;
let values = input_array.values();
// apply regexp match on the values of the array
let regexp_match_result = regexp_match(values, ..);
// form the output boolean array by looking up the result
let mut bool_builder = BooleanBuilder::new()
for key in input_array.keys() {
  bool_builder.push(regexp_match_result.value(key))
}
bool_builder.build(); // create output boolean arary
```

**Comment by alamb** [comment]
You would have to find the relevant place to plumb this in to binary.rs too I think

**Comment by alamb** [comment]
Marking as draft as I think this PR is no longer waiting on feedback. Please mark it as ready for review when it is ready for another look

**Comment by blaginin** [comment]
Thanks for the review and detailed feedback, @alamb! 😍

To give a bit of context, when submitting the PR, I was choosing between removing Regex operators from `is_comparison_operator` and manually unwrapping the dict. I chose the first option because expressions like `dict_column ~* 'scalar'` are the only outlier. `dict_column ~~ 'scalar'` and even `dict_column ~* dict_column` are both cast to UTF from dicts, so I thought consistent behavior was desirable.

However, your performance point is really valid!! Unwrapping the dict from binaries can be much more efficient than unwrapping strings and then constructing booleans... I think I've made a change; could you take a look?

A few things to consider on top of this PR:

- Maybe we should stop casting in other cases too (mentioed above)
- Feels like we might want to return results as a dict, rather than unwrapping it. I tried to make that change in this PR, but it became bigger than the actual fix 😅

File: datafusion/expr-common/src/operator.rs [hunk 1]
```diff
@@ -142,10 +142,11 @@ impl Operator {
...
     }
 
-    /// Return true if the operator is a comparison operator.
+    /// Return true if the comparison operator can be used in interval arithmetic and constraint
+    /// propagation
     ///
-    /// For example, 'Binary(a, >, b)' would be a comparison expression.
-    pub fn is_comparison_operator(&self) -> bool {
+    /// For example, 'Binary(a, >, b)' expression supports propagation.
+    pub fn supports_propagation(&self) -> bool {
```
Comment (alamb) [inline datafusion/expr-common/src/operator.rs:149]: this is technically an API change -- Can you please avoid the change and help people upgrade by adding back `is_comparison_operator` that calls `supports_propagation` and mark it deprecated?


https://github.com/apache/datafusion/blob/3353c061830dbd1ef912736fee8426ee0405a3f9/datafusion/core/src/datasource/file_format/parquet.rs#L610-L619

**Review by alamb — COMMENTED** [review]
Thank you @blaginin  -- this looks great. 

The only thing I think we should do is to deprecate `is_comparison_op` to ease people's upgrades

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
(We could add the deprecation as a follow on)

File: datafusion/expr-common/src/operator.rs [hunk 2]
```diff
@@ -164,6 +164,15 @@ impl Operator {
         )
     }
 
+    /// Return true if the comparison operator can be used in interval arithmetic and constraint
+    /// propagation
+    ///
+    /// For example, 'Binary(a, >, b)' expression supports propagation.
+    #[deprecated(since = "43.0.0", note = "please use `supports_propagation` instead")]
```
Comment (alamb) [inline datafusion/expr-common/src/operator.rs:171]: 👍

**Review by goldmedal — APPROVED** [review]
Thanks @blaginin for working on this. It looks good to me 👍

**Comment by alamb** [comment]
Thanks @blaginin  and @goldmedal for the review

**Comment by berkaysynnada** [comment]
I just noticed the change at `is_comparison_operator()`. The new name actually does not reflect what it is used for in interval arithmetic related codes. I am planning to open a PR for this. 

@blaginin do you have a preferred new name for `supports_propagation()` considering its use only in `unwrap_cast_in_comparison.rs` -- I am asking for that only since I will separate the other uses with a different name specific to constraint propagation, and narrow the scope of supported types.

**Comment by blaginin** [comment]
hey @berkaysynnada, I found the name is_comparison_operator a bit confusing because it’s not consistent with the documentation. For example:

Arc 2024-10-18 16 24 20

> do you have a preferred new name for supports_propagation() considering its use only in unwrap_cast_in_comparison.rs

Since it’s only for `unwrap_cast_in_comparison.rs`, maybe we could rename it to something like `supports_unwrap_cast`?

**Comment by berkaysynnada** [comment]
I won't revert the name change to `is_comparison_operator()`. Instead, I will create a similar function with a narrower scope, excluding operators like regex, since they aren't supported in the interval arithmetic part. In the end, there will be two functions: one for the current usage in `unwrap_cast_in_comparison.rs`, and another for interval arithmetic with a more specific set of operators. I hope this clarifies my intention.

**Comment by blaginin** [comment]
makes a lot of sense!! 👍

**Pull request merged**